File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<int> bandage, int health, vector<vector<int>> attacks) {
    int time = 0;
    int lastAttack = attacks[attacks.size()-1][0];
    int maxHealth = health;
    int attackCnt = 0;
    int 연속성공 = 0;
    while(time <= lastAttack){
        if(attacks[attackCnt][0] == time){
            연속성공 = 0;
            health -= attacks[attackCnt][1];
            attackCnt++;
            time++;
            continue;
        }
        연속성공++;
        health+=bandage[1];
        if(연속성공 == bandage[0]) {
            health+=bandage[2];
            연속성공 = 0;
        }
        if(health <= 0) return -1;
        if(health > maxHealth) health = maxHealth;
        time++;
    }
    return health <= 0 ? -1 : health;
}

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.java
class Solution {
    static int gHp;
    static int[] gBandage;
    public int solution(int[] bandage, int health, int[][] attacks) {
        gHp = health;
        gBandage = bandage;
        int beforeTime = 0;
        for(int[] attack : attacks){
            int timeDif = attack[0] - beforeTime - 1;
            beforeTime = attack[0];
            gHp += getRecovery(timeDif);
            if(gHp > health) 
                gHp = health;
            gHp -= attack[1];
            if(gHp <= 0){
                gHp = -1;
                break;
            }
        }
        return gHp;
    }
    public static int getRecovery(int timeDif){
        int recovery = 0;
        int success = timeDif/gBandage[0];
        recovery += (success * gBandage[2]);
        recovery += (gBandage[1] * timeDif);
        return recovery;
    }
}